File: experimental/math/cast_v3/tests/st/test_cast_v3.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# ----------------------------------------------------------------------------
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ----------------------------------------------------------------------------
import torch
import torch_npu  # noqa: F401


def test_cast_v3_basic():
    shapes = [(128,), (1024, 1024), (32, 64, 128), (1, 3, 224, 224)]
    dtypes = [
        (torch.float32, torch.float16),
        (torch.float16, torch.float32),
        (torch.float32, torch.int32),
        (torch.int32, torch.float32),
        (torch.int8, torch.float16),
        (torch.uint8, torch.float16),
        (torch.bool, torch.float16),
        (torch.int16, torch.float32),
        (torch.int64, torch.float32),
    ]

    for shape in shapes:
        for src_dtype, dst_dtype in dtypes:
            if src_dtype == torch.bool:
                x_cpu = torch.randint(0, 2, shape, dtype=src_dtype)
            elif src_dtype.is_floating_point:
                x_cpu = torch.randn(shape, dtype=src_dtype)
            elif src_dtype == torch.uint8:
                x_cpu = torch.randint(0, 255, shape, dtype=src_dtype)
            else:
                x_cpu = torch.randint(-128, 127, shape, dtype=src_dtype)

            x_npu = x_cpu.npu()
            y_cpu = x_cpu.to(dst_dtype)
            y_npu = x_npu.to(dst_dtype)

            if dst_dtype.is_floating_point:
                atol = 1e-3 if dst_dtype == torch.float16 else 1e-4
                rtol = 1e-3 if dst_dtype == torch.float16 else 1e-4
                ok = torch.allclose(y_npu.cpu(), y_cpu, atol=atol, rtol=rtol)
            else:
                ok = torch.equal(y_npu.cpu(), y_cpu)

            print(f"shape={shape}, {src_dtype} -> {dst_dtype}: {ok}")
            assert ok, f"Failed for shape={shape}, {src_dtype} -> {dst_dtype}"


def test_cast_v3_bf16():
    shape = (1024, 1024)
    x_fp32 = torch.randn(shape, dtype=torch.float32)
    x_npu = x_fp32.npu()

    y_bf16_npu = x_npu.to(torch.bfloat16)
    y_bf16_cpu = x_fp32.to(torch.bfloat16)
    print(
        f"fp32 -> bf16: {torch.allclose(y_bf16_npu.cpu().float(), y_bf16_cpu.float(), atol=1e-2, rtol=1e-2)}"
    )

    y_back_fp32_npu = y_bf16_npu.to(torch.float32)
    y_back_fp32_cpu = y_bf16_cpu.to(torch.float32)
    print(
        f"bf16 -> fp32: {torch.allclose(y_back_fp32_npu.cpu(), y_back_fp32_cpu, atol=1e-2, rtol=1e-2)}"
    )


def test_cast_v3_non_contiguous():
    x = torch.randn(128, 128).npu()
    x_t = x.t()
    assert not x_t.is_contiguous()
    y = x_t.to(torch.float16)
    y_cpu = x_t.cpu().to(torch.float16)
    print(
        f"non-contiguous cast: {torch.allclose(y.cpu(), y_cpu, atol=1e-3, rtol=1e-3)}"
    )


if __name__ == "__main__":
    test_cast_v3_basic()
    test_cast_v3_bf16()
    test_cast_v3_non_contiguous()
    print("All tests passed!")

File: experimental/math/cast_v3/tests/ut/op_host/op_api/test_cast_v3.cpp
/**
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */
#include <array>
#include <vector>
#include "gtest/gtest.h"

#include "aclnn_cast_v3.h"

#include "op_api_ut_common/op_api_ut.h"
#include "op_api_ut_common/tensor_desc.h"

using namespace std;

class CastV3Test : public testing::Test {
protected:
    static void SetUpTestCase() { cout << "CastV3Test SetUp" << endl; }

    static void TearDownTestCase() { cout << "CastV3Test TearDown" << endl; }
};

TEST_F(CastV3Test, case_nullptr_input)
{
    auto out_desc = TensorDesc({2, 3}, ACL_FLOAT16, ACL_FORMAT_ND);
    int64_t dstType = ACL_FLOAT16;

    auto ut = OP_API_UT(aclnnCastV3, INPUT((aclTensor*)nullptr, dstType), OUTPUT(out_desc));

    uint64_t workspace_size = 0;
    aclnnStatus aclRet = ut.TestGetWorkspaceSize(&workspace_size);
    EXPECT_EQ(aclRet, ACLNN_ERR_PARAM_NULLPTR);
}

TEST_F(CastV3Test, case_nullptr_output)
{
    auto x_desc = TensorDesc({2, 3}, ACL_FLOAT, ACL_FORMAT_ND).ValueRange(-1, 1);
    int64_t dstType = ACL_FLOAT16;

    auto ut = OP_API_UT(aclnnCastV3, INPUT(x_desc, dstType), OUTPUT((aclTensor*)nullptr));

    uint64_t workspace_size = 0;
    aclnnStatus aclRet = ut.TestGetWorkspaceSize(&workspace_size);
    EXPECT_EQ(aclRet, ACLNN_ERR_PARAM_NULLPTR);
}

TEST_F(CastV3Test, case_float_to_float16)
{
    auto x_desc = TensorDesc({2, 3, 4}, ACL_FLOAT, ACL_FORMAT_ND).ValueRange(-1, 1);
    auto out_desc = TensorDesc({2, 3, 4}, ACL_FLOAT16, ACL_FORMAT_ND);
    int64_t dstType = ACL_FLOAT16;

    auto ut = OP_API_UT(aclnnCastV3, INPUT(x_desc, dstType), OUTPUT(out_desc));

    uint64_t workspace_size = 0;
    aclnnStatus aclRet = ut.TestGetWorkspaceSize(&workspace_size);
    EXPECT_EQ(aclRet, ACL_SUCCESS);
}

TEST_F(CastV3Test, case_float16_to_float)
{
    auto x_desc = TensorDesc({1, 8}, ACL_FLOAT16, ACL_FORMAT_ND).ValueRange(-1, 1);
    auto out_desc = TensorDesc({1, 8}, ACL_FLOAT, ACL_FORMAT_ND);
    int64_t dstType = ACL_FLOAT;

    auto ut = OP_API_UT(aclnnCastV3, INPUT(x_desc, dstType), OUTPUT(out_desc));

    uint64_t workspace_size = 0;
    aclnnStatus aclRet = ut.TestGetWorkspaceSize(&workspace_size);
    EXPECT_EQ(aclRet, ACL_SUCCESS);
}

TEST_F(CastV3Test, case_float_to_int32)
{
    auto x_desc = TensorDesc({4, 5}, ACL_FLOAT, ACL_FORMAT_ND).ValueRange(-100, 100);
    auto out_desc = TensorDesc({4, 5}, ACL_INT32, ACL_FORMAT_ND);
    int64_t dstType = ACL_INT32;

    auto ut = OP_API_UT(aclnnCastV3, INPUT(x_desc, dstType), OUTPUT(out_desc));

    uint64_t workspace_size = 0;
    aclnnStatus aclRet = ut.TestGetWorkspaceSize(&workspace_size);
    EXPECT_EQ(aclRet, ACL_SUCCESS);
}

TEST_F(CastV3Test, case_int8_to_float)
{
    auto x_desc = TensorDesc({32}, ACL_INT8, ACL_FORMAT_ND).ValueRange(-128, 127);
    auto out_desc = TensorDesc({32}, ACL_FLOAT, ACL_FORMAT_ND);
    int64_t dstType = ACL_FLOAT;

    auto ut = OP_API_UT(aclnnCastV3, INPUT(x_desc, dstType), OUTPUT(out_desc));

    uint64_t workspace_size = 0;
    aclnnStatus aclRet = ut.TestGetWorkspaceSize(&workspace_size);
    EXPECT_EQ(aclRet, ACL_SUCCESS);
}

TEST_F(CastV3Test, case_uint8_to_bool)
{
    auto x_desc = TensorDesc({2, 2}, ACL_UINT8, ACL_FORMAT_ND).ValueRange(0, 255);
    auto out_desc = TensorDesc({2, 2}, ACL_BOOL, ACL_FORMAT_ND);
    int64_t dstType = ACL_BOOL;

    auto ut = OP_API_UT(aclnnCastV3, INPUT(x_desc, dstType), OUTPUT(out_desc));

    uint64_t workspace_size = 0;
    aclnnStatus aclRet = ut.TestGetWorkspaceSize(&workspace_size);
    EXPECT_EQ(aclRet, ACL_SUCCESS);
}

TEST_F(CastV3Test, case_bf16_to_float)
{
    auto x_desc = TensorDesc({16}, ACL_BF16, ACL_FORMAT_ND).ValueRange(-1, 1);
    auto out_desc = TensorDesc({16}, ACL_FLOAT, ACL_FORMAT_ND);
    int64_t dstType = ACL_FLOAT;

    auto ut = OP_API_UT(aclnnCastV3, INPUT(x_desc, dstType), OUTPUT(out_desc));

    uint64_t workspace_size = 0;
    aclnnStatus aclRet = ut.TestGetWorkspaceSize(&workspace_size);
    EXPECT_EQ(aclRet, ACL_SUCCESS);
}

TEST_F(CastV3Test, case_shape_mismatch)
{
    auto x_desc = TensorDesc({2, 3}, ACL_FLOAT, ACL_FORMAT_ND).ValueRange(-1, 1);
    auto out_desc = TensorDesc({3, 2}, ACL_FLOAT16, ACL_FORMAT_ND);
    int64_t dstType = ACL_FLOAT16;

    auto ut = OP_API_UT(aclnnCastV3, INPUT(x_desc, dstType), OUTPUT(out_desc));

    uint64_t workspace_size = 0;
    aclnnStatus aclRet = ut.TestGetWorkspaceSize(&workspace_size);
    EXPECT_EQ(aclRet, ACLNN_ERR_PARAM_INVALID);
}
